ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,888 @@
+
+  /**
+   * Determines if the offset has been copied or requires copying.
+   * @param offset the {@link Offset} to test at.
+   * @return {@code true} if copying is required. {@code false} otherwise.
+   */
+  private boolean needsCopying(Offset offset) {
+    return recoveryStartToken == null || recoveryStartToken.getOffset().compareTo(offset) < 0;
+  }
+
+  /**
+   * @return the start {@link Offset} of the index segment up until which delete records are considered applicable. Any
+   * delete records of blobs in subsequent index segments do not count and the blob is considered as not deleted.
+   * <p/>
+   * Returns {@code null} if none of the delete records are considered applicable.
+   */
+  private Offset getStartOffsetOfLastIndexSegmentForDelete() {
+    // TODO: move this to BlobStoreStats
+    long referenceTimeMs = compactionLog.getCompactionDetails().getReferenceTimeMs();
+    Offset cutoffOffset = null;
+    for (IndexSegment indexSegment : srcIndex.getIndexSegments().descendingMap().values()) {
+      if (indexSegment.getLastModifiedTimeMs() < referenceTimeMs) {
+        // NOTE: using start offset here because of the way FileSpan is treated in PersistentIndex.findKey().
+        // using this as the end offset for delete includes the whole index segment in the search.
+        cutoffOffset = indexSegment.getStartOffset();
+        break;
+      }
+    }
+    return cutoffOffset;
+  }
+
+  /**
+   * Copies data from the provided log segment into the target log (swap spaces).
+   * @param logSegmentToCopy the {@link LogSegment} to copy from.
+   * @param duplicateSearchSpan the {@link FileSpan} in which to search for duplicates.
+   * @param startOffsetOfLastIndexSegmentForDelete  the start {@link Offset} of the index segment up until which delete
+   *                                                records are considered applicable.
+   * @return {@code true} if all the records in the log segment were copied. {@code false} if some records were not
+   * copied either because there was no more capacity or because a shutdown was initiated.
+   * @throws IOException if there were I/O errors during copying.
+   * @throws StoreException if there are any problems reading or writing to store components.
+   */
+  private boolean copyDataByLogSegment(LogSegment logSegmentToCopy, FileSpan duplicateSearchSpan,
+      Offset startOffsetOfLastIndexSegmentForDelete) throws IOException, StoreException {
+    boolean allCopied = true;
+    for (Offset indexSegmentStartOffset : getIndexSegmentDetails(logSegmentToCopy.getName()).keySet()) {
+      IndexSegment indexSegmentToCopy = srcIndex.getIndexSegments().get(indexSegmentStartOffset);
+      if (needsCopying(indexSegmentToCopy.getEndOffset()) && !copyDataByIndexSegment(logSegmentToCopy,
+          indexSegmentToCopy, duplicateSearchSpan, startOffsetOfLastIndexSegmentForDelete)) {
+        // there is a shutdown in progress or there was no space to copy all entries.
+        allCopied = false;
```

Comment:
To your point though, I would argue that breaking with a flag set and then returning at the end is in no way better than returning from the middle of the method, since the control flow is the same. We have had this discussion before, and I think especially in a language like Java where exceptions are abound, the control flow argument of multiple exit points does not really hold, in the first place.